README.md
<img alt="duku" src="duku-logo.svg" width="300">

[Simple Example](#simple-example) • [Usage](#usage) • [Features](#features) • [Documentation](#documentation)

![Build Status](https://img.shields.io/github/workflow/status/oberzs/duku/Full%20Build?style=flat-square)
![Version](https://img.shields.io/badge/version-0.3.0-green?style=flat-square)
[![Rust Version](https://img.shields.io/badge/rust-1.48.0-orange?style=flat-square)](https://www.rust-lang.org/)
[![License](https://img.shields.io/github/license/oberzs/duku?style=flat-square)](https://github.com/oberzs/duku/blob/release/LICENSE)

A Rust crate for creating graphic experiences, with a focus on ease of use and beginner friendliness.
Also helpful for visualizing algorithms when learning Rust.

This creative coding library draws a lot of inspiration from [p5.js].

## Simple Example

This example draws a cube in the center of the window, rotating and coloring it based on the time that has passed.

![rotating_cube.gif](https://github.com/oberzs/duku/blob/release/examples/screenshots/rotating_cube.gif)

```rust
use duku::Camera;
use duku::Duku;
use duku::Hsb;
use duku::Light;
use duku::Result;
use std::time::Instant;

fn main() -> Result<()> {
  // create duku context and window
  let (mut duku, window) = Duku::windowed(500, 500)?;

  // create 3D camera with 90 fov
  let camera = Camera::perspective(90);

  // create directional light
  let light = Light::directional("#ffffff", [-1.0, -1.0, 1.0]);

  // start timer for rotation and color
  let timer = Instant::now();

  // start window loop
  window.while_open(move |_| {
      // start drawing on window
      duku.begin();
      duku.draw(Some(&camera), |t| {
          // setup scene
          t.background("#ababab");
          t.light(light);

          // get elapsed time since start
          let elapsed = timer.elapsed().as_secs_f32();

          // transform scene
          let angle = elapsed * 45.0;
          t.rotate_x(angle);
          t.rotate_y(angle);
          t.translate_z(2.0);

          // draw cube
          let hue = (elapsed * 60.0) as u16;
          t.tint(Hsb::new(hue, 70, 80));
          t.cube([1.0, 1.0, 1.0]);
      });
      duku.end();
  });

  Ok(())
}
```

## Usage

To use this crate, add this dependency to your `Cargo.toml` file.

```toml
[dependencies]
duku = "0.3.0"
```

## Features

- **Supports** - Windows and Linux X11
- **Vulkan** - uses the [Vulkan SDK]
- **3D** - mesh rendering with materials and shaders
- **2D** - shape, texture and text rendering with batching
- **PBR** - uses a PBR material/shader system
- **Shadows** - uses PCF shadow maps
- **Text** - uses [Fira Mono] or other OTF/TTF fonts
- **Shaders** - uses custom glsl shaders

This crate supports additional features that you can add
to your dependency in your `Cargo.toml` file.

```toml
[dependencies]
duku = { ... , features = ["feature-name"] }
```

The features include:

| Name     | Default | Uses           | Description                           |
| -------- | ------- | -------------- | ------------------------------------- |
| `window` | yes     | [winit]        | adds OS window creation support       |
| `png`    | no      | [png]          | adds png file loading support         |
| `jpeg`   | no      | [jpeg-decoder] | adds jpeg file loading support        |
| `gltf`   | no      | [gltf]         | adds gltf file loading support        |
| `glsl`   | no      | [shaderc]      | adds custom glsl file loading support |
| `otf`    | no      | [ab_glyph]     | adds otf/ttf file loading support     |
| `log`    | no      | n/a            | adds informational logs               |

## Documentation

The crate documentation can be found [here](https://docs.rs/duku) and for shader
compilation [here](https://github.com/oberzs/duku/blob/release/DC_DOCS.md). Commented examples on how to use this crate can
be found [here](https://github.com/oberzs/duku/tree/release/examples).

[p5.js]: https://p5js.org/
[crates.io]: https://crates.io
[vulkan sdk]: https://vulkan.lunarg.com/
[fira mono]: https://fonts.google.com/specimen/Fira+Mono?query=fira
[png]: https://github.com/image-rs/image-png
[jpeg-decoder]: https://github.com/image-rs/jpeg-decoder
[gltf]: https://github.com/gltf-rs/gltf
[shaderc]: https://github.com/google/shaderc-rs
[winit]: https://github.com/rust-windowing/winit
[ab_glyph]: https://github.com/alexheretic/ab-glyph


CHANGELOG.md
# Changelog

## [0.3.0] - Unreleased

### Added

- Added ability to do headless rendering.
- Added ability to save canvases to png files.
- Added explicit `begin` and `end` functions to control drawing flow.
- Added ability to get pixel data for canvases.

## [0.2.1] - 09.12.2020

### Added

- Added custom OTF/TTF font loading.

### Changed

- Use linear interpolation for text sampling.

### Fixed

- Fixed text shader for light text.

## [0.2.0] - 05.12.2020

### Added

- Added transparency blending for shapes and textures.
- Added `Send` and `Sync` traits to Handle.
- Added a `Duku` creation convenience function.
- Added `window` to default features.
- Added more conversions to `VectorN`.
- Added color gradients.
- Added per-mesh tinting.

### Changed

- Changed color API.
- Changed drawing API.
- Made autosized `Camera` the default.
- Renamed `Framebuffer` to `Canvas`.
- Renamed drawing functions.
- Renamed math structs.
- Added `Option` to some creation functions.

### Removed

- Removed `Duku::create_texture_color` function.
- Removed `Transform` struct.

### Fixed

- A lot of bugs were fixed.


src/lib.rs
// Oliver Berzs
// https://github.com/oberzs/duku

// Duku is a Vulkan rendering engine

//! This Rust crate makes it easy to render 2D and 3D graphics.
//!
//! # Example
//!
//! ```no_run
//! // This example draws a cube in the center of the window,
//! // rotating and coloring it based on the time that has passed.
//!
//! use duku::Camera;
//! use duku::Duku;
//! use duku::Hsb;
//! use duku::Light;
//! use duku::Result;
//! use std::time::Instant;
//!
//! fn main() -> Result<()> {
//!     // create duku context and window
//!     let (mut duku, window) = Duku::windowed(500, 500)?;
//!
//!     // create 3D camera with 90 fov
//!     let camera = Camera::perspective(90);
//!
//!     // create directional light
//!     let light = Light::directional("#ffffff", [-1.0, -1.0, 1.0]);
//!
//!     // start timer for rotation and color
//!     let timer = Instant::now();
//!
//!     // start window loop
//!     window.while_open(move |_| {
//!         // start drawing on window
//!         duku.draw(Some(&camera), |t| {
//!             // setup scene
//!             t.background("#ababab");
//!             t.light(light);
//!
//!             // get elapsed time since start
//!             let elapsed = timer.elapsed().as_secs_f32();
//!
//!             // transform scene
//!             let angle = elapsed * 45.0;
//!             t.rotate_x(angle);
//!             t.rotate_y(angle);
//!             t.translate_z(2.0);
//!
//!             // draw cube
//!             let hue = (elapsed * 60.0) as u16;
//!             t.tint(Hsb::new(hue, 70, 80));
//!             t.cube([1.0, 1.0, 1.0]);
//!         });
//!     });
//!
//!     Ok(())
//! }
//! ```
//!
//! More usage examples can be found [here](https://github.com/oberzs/duku/tree/release/examples).

#![warn(
    rust_2018_idioms,
    unused,
    future_incompatible,
    missing_docs,
    missing_doc_code_examples,
    unused_qualifications,
    clippy::missing_const_for_fn,
    clippy::unwrap_used,
    clippy::cast_lossless,
    clippy::clone_on_ref_ptr,
    clippy::cognitive_complexity,
    clippy::explicit_iter_loop,
    clippy::explicit_into_iter_loop,
    clippy::if_not_else,
    clippy::imprecise_flops,
    clippy::inefficient_to_string,
    clippy::unused_self
)]

// should be imported first
mod macros;

mod buffer;
mod color;
mod device;
mod duku;
mod error;
mod features;
mod font;
mod image;
mod instance;
mod math;
mod mesh;
mod pipeline;
mod renderer;
mod resources;
mod surface;
mod vk;

// normal exports
pub use self::duku::Duku;
pub use self::duku::DukuBuilder;
pub use color::Gradient;
pub use color::Hsb;
pub use color::Mix;
pub use color::Rgb;
pub use color::Rgbf;
pub use device::Stats;
pub use error::Error;
pub use error::Result;
pub use font::CharData;
pub use font::Font;
pub use font::FontData;
pub use image::Canvas;
pub use image::ColorSpace;
pub use image::Cubemap;
pub use image::CubemapSides;
pub use image::Filter;
pub use image::Format;
pub use image::Mips;
pub use image::Msaa;
pub use image::Texture;
pub use image::Wrap;
pub use math::Mat4;
pub use math::Quat;
pub use math::Vec2;
pub use math::Vec3;
pub use math::Vec4;
pub use mesh::Mesh;
pub use mesh::Model;
pub use mesh::ModelNode;
pub use pipeline::Material;
pub use pipeline::Shader;
pub use renderer::BorderMode;
pub use renderer::Camera;
pub use renderer::Light;
pub use renderer::LightType;
pub use renderer::Pcf;
pub use renderer::Projection;
pub use renderer::ShapeMode;
pub use renderer::Target;
pub use resources::Handle;
pub use resources::ReadGuard;
pub use resources::WriteGuard;
pub use surface::VSync;
pub use surface::WindowHandle;

// optional feature exports
#[cfg(feature = "glsl")]
pub use features::glsl;
#[cfg(feature = "gltf")]
pub use features::gltf;
#[cfg(feature = "otf")]
pub use features::otf;
#[cfg(feature = "window")]
pub use features::window;
